### PR TITLE
adding compress-hosts feature

### PR DESCRIPTION
**Pull request by BenCasses**
new feature : --compress-hosts or -H : compress in the other direction, search by hosts instead of attrs output has each host appearing only once.

### CODE REVIEW

File: src/libgenders/genders.h.in [hunk 1]
```diff
@@ -326,6 +326,18 @@ int genders_getattr(genders_t handle,
...
+ * genders_getflatattrsvals
+ *
+ * Writes all attrs (and "="val where present), as a single, sorted, comma
+ * delimited, null terminated string to output_string.
+ *
+ * Caller is responsible for providing an output_string long enough.
+ *
+ * Returns number of attributes on success, -1 on failure
+ */
+int genders_getflatattrsvals(genders_t gp, char *node, char *output_string);
```
Comment (chu11) [inline src/libgenders/genders.h.in]: perhaps would be safer for user to pass in length of ```output_string``` buffer, and inside function check to make sure won't buffer overflow, and if you do overflow, return appropriate error.

File: src/libgenders/genders.c [hunk 2]
```diff
@@ -818,6 +818,104 @@ genders_getattr_all(genders_t handle, char *attrs[], int len)
...
+         * I don't think that's worse than hiding them */
+        *(output_string + flatstringpos) = ',';
+        flatstringpos++;
+
+    }
+    __list_iterator_destroy( attrvallist_itr );
+
+    /* kill that last comma 
+     * also creates an empty string if no attrs */
+    *(output_string + flatstringpos - 1) = '\0';
```
Comment (chu11) [inline src/libgenders/genders.c]: is there a potential corner case if a node doesn't have any attributes here?  i.e. ```flatstringpos``` = 0

File: src/nodeattr/nodeattr.c [hunk 3]
```diff
@@ -984,10 +1004,55 @@ _hash_attrval(hash_t hattr, char *node, char *attr, char *val)
...
+
+    /* hd and hashkey are destroyed with hash_destroy(hattr) */
+     
+}
+
 static int
 _hash_hostrange(void *data, const void *key, void *arg)
 {
     struct hosts_data *hd = (struct hosts_data *)data;
+
```
Comment (chu11) [inline src/nodeattr/nodeattr.c]: errant newline?

**Comment by chu11** [comment]
overall looks good, a few general comments

1. in libgenders I believe I used GNU C style, so you may wish to emulate style there?

2. I'm not sure that ```genders_getflatattrvals()``` is truly necessary inside ```libgenders```.  Although conveniences exist inside the lib, it could probably be done outside it.  IMO, there's limited usefulness of the function outside of this one case, so I dunno if you really want to add it to the library.

3. If you keep ```genders_getflatattrvals()``` inside libgenders, remember to increase library version appropriately in ```configure.ac```.  You may also want to add tests in the testsuite and add extensions to all the appropriate parts.

File: src/nodeattr/nodeattr.c [hunk 4]
```diff
@@ -76,6 +76,8 @@ static struct option longopts[] = {
     { "diff", 1, 0, 'd'},
     { "expand", 0, 0, 'e'},
     { "compress", 0, 0, 'C'},
+    { "compress-attrs", 0, 0, 'C'},
+    { "compress-hosts", 0, 0, 'H'},
```
Comment (BenCasses) [inline src/nodeattr/nodeattr.c]: not necessary, just there for convenience as a parallel to --compress-hosts.  I'll be sure to mention it as an alias to --compress in documentation.

File: src/libgenders/genders.c (hunk 2 above)
Comment (BenCasses) [inline src/libgenders/genders.c]: yes, thank you

**Comment by BenCasses** [comment]
Seems like putting the flat string function in genders saves some work: If we rely on the current genders_getattr we'd have to recombine separate the separate attr and val char **'s, turn them into a List, sort, and then collapse to a single string.

Testing this method, the hardest genders files to process took a 20th of a second so my concern is premature optimization.  I'll make a new branch where I do all of the work in nodeattr.

**Pull request closed without merging**